### PR TITLE
Feat/migrate tsoa

### LINKED ISSUE #150: Migration to TSOA
We're migrating our project to TSOA for better API generation and maintainability. This issue serves as an umbrella for all tasks related to the migration. TSOA is a powerful tool that can help us generate robust APIs, and we want your help in making this transition.

**Tasks**:
- [x] Study the [TSOA documentation](https://github.com/lukeautry/tsoa).
- [x] Identify endpoints and controllers that need migration.
- [x] request permissions in the chat below to keep things organized. 
- [x] Convert existing controllers to TSOA controllers.
- [x] Update routing and validation as needed.
- [x] Generate and review API documentation.
- [x] Test the updated API.
- [x] Ensure existing tests are passing.

currently  contributing : 

@muttaqin1 
- [x] user_controller

@bellaabdelouahab 
- [x] auth_controller

### PR DESCRIPTION
**Pull request by muttaqin1**
This pull request will partially resolve #150.
- [x]  Migrated **admin** and **super_admin** controller to TSOA.

Any suggestions regarding the code improvement will be accepted gladly.

### CODE REVIEW

File: backend-app/controllers/users_controllers/admin_controller.ts
Comment (bellaabdelouahab) [inline backend-app/controllers/users_controllers/admin_controller.ts:125]: this will not work, I have stied it but it only shows the last one, the way i did it is like this :
```js
    @Response(
        400,
        `- This route is not for role updates. Please use /updateRole\n\- This route is not for password updates. Please use auth/updateMyPassword`
    )
    
```
it is long, I know but this is best i could do, try to update it and if you found a better solution tell me

File: backend-app/controllers/users_controllers/admin_controller.ts
Comment (bellaabdelouahab) [inline backend-app/controllers/users_controllers/admin_controller.ts:149]: In this case, you don't have to add a security decorator to each request; just add it to the controller.

File: backend-app/controllers/users_controllers/admin_controller.ts
Comment (bellaabdelouahab) [inline backend-app/controllers/users_controllers/admin_controller.ts:163]: Can you check if this is working because, for me, I do this:
```ts
public async updateMe(
        @Request() req: IReq,
        @Res() res: TsoaResponse<200, any>
    ): Promise<any> {
     //........
     return res(204, { message: 'User update successfully' });
}
```
what do you think ?

File: backend-app/controllers/users_controllers/admin_controller.ts
Comment (bellaabdelouahab) [inline backend-app/controllers/users_controllers/admin_controller.ts:170]: Remember this because you will need to change it in the next paragraph because I've created a decorator for this: @inspectAuthority("MANAGE_ROLES")

File: backend-app/controllers/users_controllers/user_controller.ts
Comment (bellaabdelouahab) [inline backend-app/controllers/users_controllers/user_controller.ts:1]: this is not in your task, I'm supposed to convert it; please drop it on your next change

File: backend-app/routes/users/admin_route.ts
Comment (bellaabdelouahab) [inline backend-app/routes/users/admin_route.ts:1]: this file will eventually be deleted

File: backend-app/utils/authorization/roles/role.ts
Comment (bellaabdelouahab) [inline backend-app/utils/authorization/roles/role.ts:5]: Can you tell me why you exported this?
because if it is needed, we can move it to interfaces instead

**Review by bellaabdelouahab — CHANGES_REQUESTED** [review]
I have given you multiple points to correct, please correct them in all files.

File: backend-app/controllers/users_controllers/admin_controller.ts
Comment (muttaqin1) [reply backend-app/controllers/users_controllers/admin_controller.ts:149]: okay.

File: backend-app/controllers/users_controllers/admin_controller.ts
Comment (muttaqin1) [reply backend-app/controllers/users_controllers/admin_controller.ts:163]: Both works.

File: backend-app/controllers/users_controllers/admin_controller.ts
Comment (muttaqin1) [reply backend-app/controllers/users_controllers/admin_controller.ts:170]: okay

File: backend-app/controllers/users_controllers/user_controller.ts
Comment (muttaqin1) [reply backend-app/controllers/users_controllers/user_controller.ts:1]: okay

File: backend-app/routes/users/admin_route.ts
Comment (muttaqin1) [reply backend-app/routes/users/admin_route.ts:1]: yes. I could delete this but i have left this task for you.

File: backend-app/utils/authorization/roles/role.ts
Comment (muttaqin1) [reply backend-app/utils/authorization/roles/role.ts:5]: I have exported this for creating dtos but this doesnt have the name prop so i have to create a new interface then i forgot to remove the export keyword. i will remove this in the next commit.

File: backend-app/controllers/users_controllers/admin_controller.ts
Comment (bellaabdelouahab) [reply backend-app/controllers/users_controllers/admin_controller.ts:163]: then it is fine but we have to choose one. By the way, btw what does 'this' refer to because it is not clear?
@muttaqin1

File: backend-app/routes/users/admin_route.ts
Comment (bellaabdelouahab) [reply backend-app/routes/users/admin_route.ts:1]: okay

File: backend-app/utils/authorization/roles/role.ts
Comment (bellaabdelouahab) [reply backend-app/utils/authorization/roles/role.ts:5]: oh, nice 👌gj

File: backend-app/controllers/users_controllers/admin_controller.ts
Comment (muttaqin1) [reply backend-app/controllers/users_controllers/admin_controller.ts:163]: In that methods context "This" is here referring to the controller class. No worries, i will go for your approach.

Comment (muttaqin1) [reply backend-app/controllers/users_controllers/admin_controller.ts:163]: This class has some properties called 
```js
Class someContoller{
constructor(statusCode){
this.statusCode= statusCode;
}

setStatus(statusCode){
this.statusCode= statusCode;
}

Login(){
this.setStatus(200);
return {}//some response data
}
/*so when this login method runs the status code is set to 200 for this method. The controller takes the current status code and send response with this status code. This **statusCode** will be modified for other methods.
*/
//this is just my assumption.
}
```

Comment (muttaqin1) [reply backend-app/controllers/users_controllers/admin_controller.ts:163]: @bellaabdelouahab

File: backend-app/controllers/users_controllers/admin_controller.ts
Comment (muttaqin1) [reply backend-app/controllers/users_controllers/admin_controller.ts:125]: okay

File: backend-app/controllers/users_controllers/admin_controller.ts
Comment (bellaabdelouahab) [reply backend-app/controllers/users_controllers/admin_controller.ts:163]: but what about if the class has 100 request each one is changing the 'this' hhhhh ?

Comment (muttaqin1) [reply backend-app/controllers/users_controllers/admin_controller.ts:163]: You will be shocked to know that behind the scene   **res** also does the same.

Comment (bellaabdelouahab) [reply backend-app/controllers/users_controllers/admin_controller.ts:163]: Oh, really, if that's the case, then we can keep both but only for code clarity 
@muttaqin1

**Comment by bellaabdelouahab** [comment]
okay, should I wait for this to be merged or submit another pr

File: backend-app/controllers/users_controllers/admin_controller.ts
Comment (muttaqin1) [reply backend-app/controllers/users_controllers/admin_controller.ts:163]: No, i have already told you that i am going to follow your approach. Your approach is very clear and the code looks well👌

**Comment by muttaqin1** [comment]
> okay, should I wait for this to be merged or submit another pr

Wait. I will complete this tomorrow insh allah.

**Comment by muttaqin1** [comment]
@bellaabdelouahab

File: backend-app/controllers/users_controllers/admin_controller.ts
Comment (bellaabdelouahab) [inline backend-app/controllers/users_controllers/admin_controller.ts]: where did the 404 response go?

File: backend-app/controllers/users_controllers/admin_controller.ts
Comment (bellaabdelouahab) [inline backend-app/controllers/users_controllers/admin_controller.ts:166]: We will do another review to remove unnecessary 'any' keywords.

**Review by bellaabdelouahab — CHANGES_REQUESTED** [review]
here is my review .
note: you havent revoke the user_controller.ts ?

**Review by bellaabdelouahab — APPROVED** [review]
Thank you that is all, there are some small changes but i will handle them

**Comment by muttaqin1** [comment]
> here is my review .
> note: you havent revoke the user_controller.ts ? 

But i have deleted that commit then force pushed the changes.
@bellaabdelouahab

**Comment by bellaabdelouahab** [comment]
> But i have deleted that commit then force pushed the changes.
> @bellaabdelouahab 
But that is not how to fix it you deleted the commit but the changes are still on other commits after it 
The way to do it is revok changes on the file

**Comment by bellaabdelouahab** [comment]
@muttaqin1

**Comment by bellaabdelouahab** [comment]
I will push mine after i merge @muttaqin1

**Comment by muttaqin1** [comment]
> I will push mine after i merge @muttaqin1 

Okay

**Pull request merged**
